snewpdag/plugins/Chi2Calculation.py

```diff
@@ -0,0 +1,158 @@
+        return np.linalg.inv(V)
+
+    # Generates unit vector for given lattitude and longnitude,
+    # pointing towards sky
+    # alpha range is (-pi, pi), delta range is (-pi/2, pi/2)
+    def angles_to_unit_vec(self, lon, lat):
+        x = np.cos(lon)*np.cos(lat)
+        y = np.sin(lon)*np.cos(lat)
+        z = np.sin(lat)
+        return np.matrix([x, y, z]).getT()
+
+    # Calculates detecotr position in cartesian coordinates
+    def det_cartesian_position(self, det):
+        ang_rot = 7.29e-5  # radians/s
+        ang_sun = 2e-7  # radians/s   2pi/365days
+
+        # take into account the time dependence of longitude
+        # reference: arXiv:1304.5006
+        arrival_date = datetime.fromtimestamp(self.arrival[0])
+        decimal = self.arrival[1]*1e-9
+
+        t_rot = arrival_date.hour*60*60 \
+              + arrival_date.minute*60 + arrival_date.second + decimal
+
+        t_sun = self.arrival[0] - 953582400 + decimal
+
+        lon = det[0] + ang_rot*t_rot - ang_sun*t_sun - np.pi
+        lat = det[1]
+        r = 6.37e6 + det[2]
+
+        return r*self.angles_to_unit_vec(lon, lat)
+
+    # Calculates time_diff given detector names and supernova location
+    def time_diff(self, det1, det2, n):
+        c = 3.0e8  # speed of light /m*s^-1
+
+        det1_pos = self.det_cartesian_position(det1)
+        det2_pos = self.det_cartesian_position(det2)
+
+        diff = float((det1_pos - det2_pos).getT() @ n)/c
+
+        return diff
+
+    # Calculates chi2 for given vector d
+    def chi2(self, d):
+        return (d.getT() @ (self.precision_matrix @ d))
+
+    # Calculates vector d given supernova position and time differences
+    def d_vec(self, n, dt_i):
+        n_detectors = len(self.detector_info)
+        d = np.zeros(n_detectors)
+
+        for i in range(n_detectors):
+            det = list(self.detector_info.keys())[i]
+            det0 = self.d0_info[-1]
+            d[i] = dt_i[det][0] + dt_i[det][1] / 1e9 \
+                 - dt_i[det0][0] - dt_i[det0][1] / 1e9
+
+            d[i] = d[i] - self.detector_info[det][4] + self.d0_info[4]
+            d[i] -= self.time_diff(self.detector_info[det], self.d0_info, n)
+
+        return np.matrix(d).getT()
+
+    # Generates chi2 map
+    def generate_map(self, dt_i):
+        map = np.zeros(self.NPIX)
+
+        for i in range(self.NPIX):
+            delta, alpha = hp.pixelfunc.pix2ang(self.NSIDE, i, nest=True, lonlat=True)
+
+            n_pointing = -1*self.angles_to_unit_vec(alpha, delta)
+            map[i] = self.chi2(self.d_vec(n_pointing, dt_i))
+
+        chi2_min = map.min()
+        for i in range(self.NPIX):
+            map[i] -= chi2_min
+
+        return map
+
+    def alert(self, data):
+        if self.ndof < 1:
+            logging.error("Not enough data for chi2 map calculation")
+            return False
+
+        dt_i = data['gen']['neutrino_times']
+        self.arrival = data['gen']['sn_times']['Earth']
```

Comment:
this works for analyzing the MC input. Data will be a little different (will try to describe in more detail elsewhere)